### PR TITLE
Look at direction of analog axis travel instead of instantaneous sample

### PR DESCRIPTION
**Pull request by xperia64**
When mapping an analog trigger to a button, take two samples of an axis to determine the direction it's being moved, and override the event value to be saturated in that direction if it's been moved enough to count as an event.

Should improve https://github.com/citra-emu/citra/issues/4166 and https://github.com/citra-emu/citra/issues/5488, and possibly https://github.com/citra-emu/citra/issues/3791 by making the controller mapping process less confusing.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5509)

### CODE REVIEW

File: src/input_common/sdl/sdl_impl.cpp [hunk 1]
```diff
@@ -594,8 +594,17 @@ class SDLButtonPoller final : public SDLPoller {
...
-                if (std::abs(event.jaxis.value / 32767.0) < 0.5) {
+                if (axis_memory.find(event.jaxis.axis) == axis_memory.end())
+                {
+                    axis_memory[event.jaxis.axis] = event.jaxis.value;
                     break;
+                } else {
+                    if (std::abs((event.jaxis.value - axis_memory[event.jaxis.axis]) / 32767.0) < 0.5) {
+                        break;
+                    } else {
+                        event.jaxis.value = std::copysign(32767, event.jaxis.value - axis_memory[event.jaxis.axis]);
```
Comment (vitor-k) [inline src/input_common/sdl/sdl_impl.cpp:605]:
```suggestion
                        event.jaxis.value =
                            std::copysign(32767, event.jaxis.value - axis_memory[event.jaxis.axis]);
```

File: src/input_common/sdl/sdl_impl.cpp [hunk 2]
```diff
@@ -594,8 +594,17 @@ class SDLButtonPoller final : public SDLPoller {
         while (state.event_queue.Pop(event)) {
             switch (event.type) {
             case SDL_JOYAXISMOTION:
-                if (std::abs(event.jaxis.value / 32767.0) < 0.5) {
+                if (axis_memory.find(event.jaxis.axis) == axis_memory.end())
+                {
```
Comment (vitor-k) [inline src/input_common/sdl/sdl_impl.cpp:598]:
```suggestion
                if (axis_memory.find(event.jaxis.axis) == axis_memory.end()) {
```

File: src/input_common/sdl/sdl_impl.cpp [hunk 3]
```diff
@@ -594,8 +594,17 @@ class SDLButtonPoller final : public SDLPoller {
...
+                    axis_memory[event.jaxis.axis] = event.jaxis.value;
                     break;
+                } else {
+                    if (std::abs((event.jaxis.value - axis_memory[event.jaxis.axis]) / 32767.0) < 0.5) {
```
Comment (vitor-k) [inline src/input_common/sdl/sdl_impl.cpp:602]:
```suggestion
                    if (std::abs((event.jaxis.value - axis_memory[event.jaxis.axis]) / 32767.0) <
                        0.5) {
```

**Comment by xperia64** [comment]
This code also makes https://github.com/citra-emu/citra/pull/4984 obsolete and handles the case mentioned in the comments.

**Review by B3n30 — APPROVED** [review]

File: src/input_common/sdl/sdl_impl.cpp [hunk 4]
```diff
@@ -594,8 +594,18 @@ class SDLButtonPoller final : public SDLPoller {
         while (state.event_queue.Pop(event)) {
             switch (event.type) {
             case SDL_JOYAXISMOTION:
-                if (std::abs(event.jaxis.value / 32767.0) < 0.5) {
+                if (axis_memory.find(event.jaxis.axis) == axis_memory.end()) {
```
Comment (zhaowenlan1779) [inline src/input_common/sdl/sdl_impl.cpp:597]:
```suggestion
                if (!axis_memory.count(event.jaxis.axis)) {
```
And invert the clauses.

File: src/input_common/sdl/sdl_impl.cpp [hunk 5]
```diff
@@ -604,6 +614,9 @@ class SDLButtonPoller final : public SDLPoller {
         }
         return {};
     }
+
+private:
+    std::map<int, int> axis_memory;
```
Comment (zhaowenlan1779) [inline src/input_common/sdl/sdl_impl.cpp:619]: `<map>` should be included.

Comment (BreadFish64) [reply src/input_common/sdl/sdl_impl.cpp:619]: I'd use unordered_map here as well.

**Comment by Morph1984** [comment]
This is a candidate to be ported over to yuzu

**Comment by xperia64** [comment]
While this improves analog axis mappings for true analog triggers, it causes problems for controllers with digital buttons mapped to an analog axis (e.g. my 3rd party switch controller which shows up as XInput on PC). They send one event on press and one event on release, and this code requires two events to complete a mapping. Going to rework it.

**Comment by xperia64** [comment]
I've improved the heuristics and have tested a variety of controllers:
Normal controllers with true analog triggers/sticks (Xbox One, DS4)
Switch Pro clone controller with digital LT/RT triggers
Sega Genesis Mini controller with a digital D-Pad that's mapped to analog axes for some reason


I'm sure there are other controllers that may break these heuristics, but it's working far better than it was before IMO.

The one edge case I've found is if a user presses true analog triggers too quickly OR has a very slow update rate for some reason, it may incorrectly consider them as digital.

**Comment by xperia64** [comment]
Since this got merged into Yuzu, pulled the trigger here.

**Pull request merged**